src/lib/zkstring.ts

```diff
@@ -0,0 +1,148 @@
+import { Bool, Circuit, Field, Poseidon } from '../snarky';
+import { CircuitValue, prop } from './circuit_value';
+
+export class Character extends CircuitValue {
+  @prop value: Field;
+
+  constructor(value: Field) {
+    super();
+
+    if (Character.validateInput(value)) {
+      this.value = value;
+    } else {
+      throw Error('Character initialization did not pass validation');
+    }
+  }
+
+  equals(char: Character): Bool {
+    return this.value.equals(char.value);
+  }
+
+  toField(): Field {
+    return this.value;
+  }
+
+  toString(): string {
+    const charCode = Number(this.value.toString());
+    return String.fromCharCode(charCode);
+  }
+
+  static fromString(str: string) {
+    const char = Field(str.charCodeAt(0));
+    return new Character(char);
+  }
+
+  /* 
+  TODO: Add support for more character sets
+  ASCII reference: https://en.wikipedia.org/wiki/ASCII#Printable_characters
+  */
+  static validateInput(value: Field): Boolean {
```

Comment:
This function should be called `check`, then it will run automatically when a witness is created, so ZkStrings that are passed to a method (for example) will be constrained inside the circuit

Reply:
Also, let's remove the check for inProver / inCheckedComputation, if we don't call it implicitly in the constructor as suggested above

Reply:
so `check` is a magic method name?  Are there others?

Reply:
no there aren't, at least for CircuitValue 😅